File: src/main/scala/com/reroute/backend/stations/transitland/TransitlandApi.java
package com.reroute.backend.stations.transitland;

import com.reroute.backend.model.distance.Distance;
import com.reroute.backend.model.location.LocationPoint;
import com.reroute.backend.utils.LoggingUtils;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import org.json.JSONArray;
import org.json.JSONObject;

import java.net.URL;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.concurrent.TimeUnit;

/**
 * Created by ilan on 1/22/17.
 */
public class TransitlandApi {

    private static final String BASE_FEED_URL = "http://transit.land/api/v1/feeds";
    private static final String AREA_FORMAT = "bbox=%f,%f,%f,%f";

    private static final double MILES_TO_LAT = 1.0 / 70;
    private static final double MILES_TO_LONG = 1.0 / 70;


    private JSONObject callUrl(String url) {
        OkHttpClient client;
        client = new OkHttpClient.Builder()
                .connectTimeout(120, TimeUnit.SECONDS)
                .readTimeout(120, TimeUnit.SECONDS)
                .writeTimeout(120, TimeUnit.SECONDS)
                .build();


        Request request = new Request.Builder()
                .url(url)
                .build();

        Response response;
        JSONObject rawobj;
        try {
            response = client.newCall(request).execute();
            if (!response.isSuccessful()) {
                return null;
            }
            rawobj = new JSONObject(response.body().string());
        } catch (Exception e) {
            LoggingUtils.logError(e);
            return null;
        }

        return rawobj;
    }

    public List<URL> getFeedsInArea(LocationPoint center, Distance range, Map<String, String> restrict, Map<String, String> avoid) {
        String url = BASE_FEED_URL + "?per_page=500";
        if (center != null && range != null && range.inMeters() >= 0) {
            double latVal1 = center.getCoordinates()[0] - range.inMiles() * MILES_TO_LAT;
            double latVal2 = center.getCoordinates()[0] + range.inMiles() * MILES_TO_LAT;

            double lngVal1 = center.getCoordinates()[1] - range.inMiles() * MILES_TO_LONG;
            double lngVal2 = center.getCoordinates()[1] + range.inMiles() * MILES_TO_LONG;
            url += "&" + String.format(AREA_FORMAT, lngVal1, latVal1, lngVal2, latVal2);
        }

        try {
            JSONObject obj = callUrl(url);
            if (obj == null) return Collections.emptyList();

            JSONArray feeds = obj.getJSONArray("feeds");
            int len = feeds.length();
            List<URL> rval = new ArrayList<>(len);
            for (int i = 0; i < len; i++) {
                JSONObject curobj = feeds.getJSONObject(i);
                System.out.println(curobj.toString(1));
                boolean works = (restrict == null || restrict.keySet()
                        .stream()
                        .allMatch(key -> curobj.get(key).equals(restrict.get(key))))
                        && (avoid == null || avoid.keySet()
                        .stream()
                        .noneMatch(key -> curobj.get(key).equals(avoid.get(key))));
                if (works) rval.add(new URL(curobj.getString("url")));
            }
            return rval;
        } catch (Exception e) {
            LoggingUtils.logError(e);
            return Collections.emptyList();
        }
    }

}

File: src/main/scala/com/reroute/backend/stations/transitland/TransitlandApi.scala
package com.reroute.backend.stations.transitland

import java.net.URL
import java.util
import java.util.Collections
import java.util.concurrent.TimeUnit

import com.reroute.backend.model.distance.{DistUnits, Distance}
import com.reroute.backend.model.location.LocationPoint
import okhttp3.{OkHttpClient, Request}
import org.json.JSONObject

/**
 * Created by ilan on 1/22/17.
 */
object TransitlandApi {
  private val BASE_FEED_URL = "http://transit.land/api/v1/feeds"
  private val MILES_TO_LAT = 1.0 / 70
  private val MILES_TO_LONG = 1.0 / 70
}

class TransitlandApi {
  private def callUrl(url: String): JSONObject = {
    val client = new OkHttpClient.Builder()
      .connectTimeout(120, TimeUnit.SECONDS)
      .readTimeout(120, TimeUnit.SECONDS)
      .writeTimeout(120, TimeUnit.SECONDS)
      .build()
    val request = new Request.Builder().url(url).build
    val response = client.newCall(request).execute
    if (!response.isSuccessful) return new JSONObject()
    new JSONObject(response.body.string)
  }

  def getFeedsInArea(center: LocationPoint, range: Distance, restrict: util.Map[String, String],
                     avoid: util.Map[String, String]): util.List[URL] = {
    var url = TransitlandApi.BASE_FEED_URL + "?per_page=500"
    if (center != null && range != null && range.in(DistUnits.METERS) >= 0) {
      val latVal1 = center.latitude - range.in(DistUnits.MILES) * TransitlandApi.MILES_TO_LAT
      val latVal2 = center.latitude + range.in(DistUnits.MILES) * TransitlandApi.MILES_TO_LAT
      val lngVal1 = center.longitude - range.in(DistUnits.MILES) * TransitlandApi.MILES_TO_LONG
      val lngVal2 = center.longitude + range.in(DistUnits.MILES) * TransitlandApi.MILES_TO_LONG
      url = s"$url&bbox=$lngVal1,$latVal1,$lngVal2,$latVal2"
    }
    val obj = callUrl(url)
    if (obj.keySet().isEmpty) return Collections.emptyList()
    val feeds = obj.getJSONArray("feeds")
    val len = feeds.length
    val rval = new util.ArrayList[URL](len)
    (0 until len)
      .map(feeds.getJSONObject)
      .filter(meetsFilter(restrict, avoid))
      .map(obj => new URL(obj.getString("url")))
      .foreach(rval.add)
    rval
  }

  def meetsFilter(restrict: util.Map[String, String],
                  avoid: util.Map[String, String]): JSONObject => Boolean = curobj => {
    val restrictEmpty = restrict == null || restrict.isEmpty
    val restrictWorks = restrictEmpty || restrict.keySet.stream.allMatch(key => curobj.get(key) == restrict.get(key))

    val avoidEmpty = avoid == null || avoid.isEmpty
    val avoidWorks = avoidEmpty || avoid.keySet.stream.noneMatch(key => curobj.get(key) == avoid.get(key))

    restrictWorks && avoidWorks
  }
}